transform_aes.c

```diff
@@ -219,28 +220,37 @@ static int transop_encode_aes( n2n_trans_op_t * arg,
             /* Encode the security association (SA) number */
             encode_uint32( outbuf, &idx, sa->sa_id );
 
-            /* Encrypt the assembly contents and write the ciphertext after the SA. */
-            len = in_len + TRANSOP_AES_NONCE_SIZE;
+           /* The following fields (message authentication, IV) are calculated respectively filled after data encrpytion */
 
-            /* The assembly buffer is a source for encrypting data. The nonce is
+            /* Encrypt the assembly contents and write the ciphertext after the IV. */
+            len = in_len + TRANSOP_AES_DATA_LEN;
+
+            /* The assembly buffer is a source for encrypting data. The data length is
              * written in first followed by the packet payload. The whole
              * contents of assembly are encrypted. */
-            pnonce = (uint32_t *)assembly;
-            *pnonce = rand();
-            memcpy( assembly + TRANSOP_AES_NONCE_SIZE, inbuf, in_len );
+            data_len = (uint16_t *)assembly;
+            *data_len = in_len;
 
-            /* Need at least one encrypted byte at the end for the padding. */
-            len2 = ( (len / AES_BLOCK_SIZE) + 1) * AES_BLOCK_SIZE; /* Round up to next whole AES adding at least one byte. */
-            assembly[ len2-1 ]=(len2-len);
-            traceEvent( TRACE_DEBUG, "padding = %u", assembly[ len2-1 ] );
+	    memcpy( assembly + TRANSOP_AES_DATA_LEN , inbuf, in_len );
+            len2 = ( ( ( len - 1) / AES_BLOCK_SIZE) + 1 ) * AES_BLOCK_SIZE; /* Round up to next whole AES block size */
+
+            RAND_bytes(&iv, 16);  /* get a cryptographically safely randomized initialization vector */
+            /* REVISIT: return value of RAND_bytes function should be 1 to continue cryptographically safe, otherwise exit */
+            memcpy((sa->enc_ivec), iv, 16 ); /* sa->enc_ivec gets changed during CBC encryption procedure, so keep a copy in iv */
```

Comment:
Is it necessary to provide the IV on every packet? In my understanding, the IV is only required during the decryption of the very first block, is this correct?

Reply:
The IV is different for each packet. So, it must be provided once for every transmitted packet - at the beginning of the CBC decryption process.

It is generated and stored in "iv". Afterwards, it gets copied to enc_ivec. During the CBC decrpytion process, it seems that the content of enc_ivec gets changed with every single block inside the CBC crypto function. I guess this from some simple debugging showing that enc_ivec has a different content after the CBC function call. That's why we should keep it somewhere, here iv to use it later to put in the packet.

Ummm, maybe I could have directly generated it into the packet buffer and afterwards copy it into enc_ivec? How could byte order be correct?

Reply:
For Part 2 I already took a note to remove enc_iv an dec_iv from the sa structure and use local variables only. As IVs are varying for every packet, there is no need to keep track of them in the more static sa structure.